### DISCUSSION TITLE
Adding kubernetes labels to Kepler metrics

### DISCUSSION THREAD

**Discussion by AdrianHammond** [sustainable-computing-io/kepler#756] [General]
Does Kepler, or can Kepler, include K8s label with the data it exports. The use case is that labels are used to tag namespaces / project with organisation data to make it easier to attribute ownership. 

Thanks Adrian

**Comment by rootfs** [comment]
kepler metrics currently have pod name and pod namespace. It wouldn't be hard to group/aggregate/filter by [kube_pod_labels](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/pod-metrics.md)

**Reply by AdrianHammond** [reply]
Would it help for me to get more info on what other labels the customer uses for you? I assume this would be an enhancement to Kepler?

**Reply by rootfs** [reply]
we can start with prometheus query, the [join metrics](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/README.md#join-metrics) example could be a good start.

**Comment by AdrianHammond** [comment]
@rootfs  - I spoke to the customer who raised this with me today, they use n/s labels to associate all applications with a business unit. They want to be able to visualise data at the BU level in addition to individual n/s. Thanks A

**Comment by AdrianHammond** [comment]
This JIRA has been opened - https://issues.redhat.com/browse/OBSDA-496

**Comment by marceloamaral** [comment]
First draft implementation of this: https://github.com/marceloamaral/label-exporter

**Reply by rootfs** [reply]
thank you @marceloamaral 
@AdrianHammond can you take a look and see if this works for your case?

**Reply by AdrianHammond** [reply]
Thanks for progressing this @marceloamaral 

The customer request was specifically for namespace labels, as label_exporter includes pod_namespace that should be okay.  Will the customer need to change there labelling scheme and have to add `le_` to labels they want export? If so that might be an issue as that would require them to change their labelling scheme and update existing label or add new labels.

@rootfs

**Reply by marceloamaral** [reply]
Kepler metrics currently have pod namespace labels, so you probably don't need the label-exporter.

The label-exporter is to expose additional custom labels. For example `group=1` so that all pods related to group1 can be aggregated.
The necessity of a prefix in the labels `le_` is to minimize the cardinality of prometheus metrics which can become a scalability problem. But it is something that can be made more flexible.

**Reply by AdrianHammond** [reply]
@marceloamaral  - just to clarify what I should have said. The customer use case is to aggregate kepler metrics based on additional namespace labels - not just the namespace name. For example `application.name/retail` which could be a label on multiple namespaces being used by the _retail_ application.

**Reply by marceloamaral** [reply]
Does the use need more than one label?
If only one label is needed, you can just change the prefix `--label-prefix="application.name"`.
If the user need more labels, I change update it to the `label-prefix` parameter be a list.

**Reply by AdrianHammond** [reply]
Yes, likely they would use multiples. e.g. `application.name/retail` and `application.process/payment`

**Reply by marceloamaral** [reply]
@AdrianHammond I will update for the list then!
Also, I updated the README so you can see the example of using only one label for now.

**Reply by marceloamaral** [reply]
@AdrianHammond, updated the pod label exporter. The user can configure it with a list of labels now. 
You can see the example in the github README.

**Reply by AdrianHammond** [reply]
@marceloamaral LGTM - thank you

**Comment by sthaha** [comment]
@rootfs @marceloamaral , Could we consider allowing the same functionality as a feature flag in kepler? e.g. 
`kepler --collector.labels=<label-key> --collector.labels=<label-key>` this would allows kepler to export labels only if required and the overhead should be minimal. The additional benefit is that it allows for the deployment management (daemonsets, rbac, configuration, ) and project management (release and maintenance  etc) to be minimal.

**Reply by marceloamaral** [reply]
The label exporter should be a single deployment in the cluster, and then we can also support other kind of objects such as Job. Don't need to be only pods.

Note that we cannot watch for jobs in Kepler because it will introduce too much overhead in the API server as Jobs cannot be filtered per node as we can do for pods.

Also, Kepler metrics already have many labels and we should not increase too much the cardinality because of Prometheus overhead constraints.

@rootfs can we move the `label-exporter` under the `https://github.com/sustainable-computing-io/`?
Once we have it there, we could support the deployment via operator.

**Reply by sthaha** [reply]
I agree on the cardinality in general but  in my humble opinion that choice can be left to users? By default, kepler collects no additional labels. Also, I believe k-s-m already has the feature to collect pod labels see: https://github.com/kubernetes/kube-state-metrics/blob/main/docs/pod-metrics.md. But again, you don't need an additional deployment/  overhead in cluster if kepler provides this (optional) feature.

**Reply by rootfs** [reply]
thank @marceloamaral the label-exporter is now here https://github.com/sustainable-computing-io/label-exporter
